### PR TITLE
Make minio package support legacy MD5 checksum

### PR DESCRIPTION
**Pull request by wxiaoguang**
A feedback from discord: https://discord.com/channels/322538954119184384/561007778139734027/1090185427115319386

Some storages like:

 * https://developers.cloudflare.com/r2/api/s3/api/
 * https://www.backblaze.com/b2/docs/s3_compatible_api.html

They do not support "x-amz-checksum-algorithm" header

But minio recently uses that header with CRC32C by default. So we have to tell minio to use legacy MD5 checksum.

I guess this needs to be backported because IIRC we 1.19 and 1.20 are using similar minio package.

The minio package code for SendContentMD5 looks like this:

<details>

image

</details>

### CODE REVIEW

File: modules/storage/minio.go [hunk 1]
```diff
@@ -142,7 +142,14 @@ func (m *MinioStorage) Save(path string, r io.Reader, size int64) (int64, error)
 		m.buildMinioPath(path),
 		r,
 		size,
-		minio.PutObjectOptions{ContentType: "application/octet-stream"},
+		minio.PutObjectOptions{
+			ContentType: "application/octet-stream",
+			// some storages like:
+			// * https://developers.cloudflare.com/r2/api/s3/api/
+			// * https://www.backblaze.com/b2/docs/s3_compatible_api.html
+			// they do not support "x-amz-checksum-algorithm" header, so use legacy MD5 checksum
```
Comment (yardenshoham) [inline modules/storage/minio.go]:
```suggestion
			// do not support "x-amz-checksum-algorithm" header, so use legacy MD5 checksum
```

**Comment by KN4CK3R** [comment]
Should we make that configurable? That setting feels like "legacy" and may prevent working with newer services.

**Comment by lunny** [comment]
> @KN4CK3R I found a low-level bug in MinIO package, it ignores some errors if the written side could match the expected size. The bug only happens when `SendContentMD5=true`
> 
> So I have to fix the bug on our side , see [dc7fc8b](https://github.com/go-gitea/gitea/commit/dc7fc8b1933817d118a868d237938f05b005b6ca) , `recordError`

Could we upgrade our minio sdk to test if it's fixed?

**Review by lunny — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

File: modules/lfs/content_store.go [hunk 2]
```diff
@@ -109,6 +115,17 @@ type hashingReader struct {
...
+}
+
+// recordError records the last error during the Save operation
+// Some callers of the Reader doesn't respect the returned "err"
+// For example, MinIO's Put will ignore errors if the written size could equal to expected size
+// So we must remember the error by ourselves,
+// and later check again whether ErrSizeMismatch or ErrHashMismatch occurs during the Save operation
+func (r *hashingReader) recordError(n int, err error) (int, error) {
+	r.lastError = err
+	return n, err
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]:
```suggestion
func (r *hashingReader) recordError(err error) (error) {
	r.lastError = err
	return err
```

File: modules/lfs/content_store.go [hunk 3]
```diff
@@ -118,22 +135,22 @@ func (r *hashingReader) Read(b []byte) (int, error) {
 		r.currentSize += int64(n)
 		wn, werr := r.hash.Write(b[:n])
 		if wn != n || werr != nil {
-			return n, werr
+			return r.recordError(n, werr)
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]:
```suggestion
			return n, r.recordError(werr)
```

File: modules/lfs/content_store.go [hunk 4]
```diff
@@ -118,22 +135,22 @@ func (r *hashingReader) Read(b []byte) (int, error) {
...
 		}
 	}
 
-	if err != nil && err == io.EOF {
+	if errors.Is(err, io.EOF) || r.currentSize >= r.expectedSize {
 		if r.currentSize != r.expectedSize {
-			return n, ErrSizeMismatch
+			return r.recordError(n, ErrSizeMismatch)
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]:
```suggestion
			return n, r.recordError(ErrSizeMismatch)
```

File: modules/lfs/content_store.go [hunk 5]
```diff
@@ -118,22 +135,22 @@ func (r *hashingReader) Read(b []byte) (int, error) {
...
 		}
 
 		shaStr := hex.EncodeToString(r.hash.Sum(nil))
 		if shaStr != r.expectedHash {
-			return n, ErrHashMismatch
+			return r.recordError(n, ErrHashMismatch)
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]:
```suggestion
			return n, r.recordError(ErrHashMismatch)
```

File: modules/lfs/content_store.go [hunk 6]
```diff
@@ -118,22 +135,22 @@ func (r *hashingReader) Read(b []byte) (int, error) {
...
 		}
 	}
 
-	return n, err
+	return r.recordError(n, err)
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]:
```suggestion
	return n, r.recordError(err)
```

**Review by KN4CK3R — CHANGES_REQUESTED** [review]

File: modules/lfs/content_store.go [hunk 7]
```diff
@@ -60,6 +60,12 @@ func (s *ContentStore) Put(pointer Pointer, r io.Reader) error {
 		return err
 	}
 
+	// check again whether there is any error during the Save operation
+	// because some errors might be ignored by the Reader's caller
+	if wrappedRd.lastError != nil && !errors.Is(wrappedRd.lastError, io.EOF) {
+		return wrappedRd.lastError
```
Comment (KN4CK3R) [inline modules/lfs/content_store.go]: If `err == nil` and `wrappedRd.lastError != nil` the file was uploaded and needs to get removed. Currently only the error is returned but the file is still there.

**Comment by KN4CK3R** [comment]
Here is the code which drops the error btw: https://github.com/minio/minio-go/blob/f54fa21d32c040f6b33ac949b72c58bcea5d0669/utils.go#L389-L422

**Review by KN4CK3R — APPROVED** [review]

**Comment by douglasparker** [comment]
Thank you so much! This was the missing piece for me to migrate my projects to Gitea! 🥳

**Pull request merged**